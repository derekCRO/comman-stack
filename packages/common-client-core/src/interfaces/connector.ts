
export interface ClientStateParams {
    resolvers: any;
    defaults?: any;
    typeDefs?: string | string[];
}

export interface FeatureParams {
    readonly link?: any;
    readonly createFetch?: any;
    readonly connectionParam?: any;
    readonly reducer?: any;
    readonly resolver?: any;
    readonly routerFactory?: any;
    readonly route?: any;
    readonly routeConfig?: any;
    readonly navItem?: any;
    readonly navItemRight?: any;
    readonly rootComponentFactory?: any;
    readonly dataRootComponent?: any;
    readonly createFetchOptions?: any;
    readonly stylesInsert?: any;
    readonly scriptsInsert?: any;
    readonly catalogInfo?: any;
    readonly languagesFuncs?: any;
}


export interface IFeature {
    // Public variables
    readonly link: any;
    readonly createFetch: any;
    readonly connectionParam: any;
    readonly reducer: any;
    readonly resolver: any;
    readonly routerFactory: any;
    readonly route: any;
    readonly routeConfig: any;
    readonly navItem: any;
    readonly navItemRight: any;
    readonly rootComponentFactory: any[];
    readonly dataRootComponent: any[];
    readonly createFetchOptions: any[];
    readonly stylesInsert: any[];
    readonly scriptsInsert: any[];
    readonly catalogInfo: any[];
    readonly languagesFuncs: any[];
    readonly data: any[];

    // methods
    readonly router;

    readonly routes;
    readonly configuredRoutes;
    readonly navItems;
    readonly navItemsRight;

    readonly reducers;
    readonly resolvers;

    readonly connectionParams;

    readonly stylesInserts;
    readonly scriptsInserts;

    getWrappedRoot(root: any, req?: any): any;
    getDataRoot(root: any): any;
    registerLanguages(monaco): void;
}
