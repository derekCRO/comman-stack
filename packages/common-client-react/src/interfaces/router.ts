import { LoadingComponentProps } from 'react-loadable';

export interface IRouteData {
    [key: string]: {
        component: () => any | any;
        loading?: React.ComponentType<LoadingComponentProps> | (() => null);
    };
}
