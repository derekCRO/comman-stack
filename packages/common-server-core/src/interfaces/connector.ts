import * as Logger from 'bunyan';
import { PubSub } from 'graphql-subscriptions';

export interface IResolverOptions {
    pubsub: any;
    logger?: Logger;
    [key: string]: any;
}

export interface IDirectiveOptions {
    [key: string]: any;
    logger?: Logger;
}
