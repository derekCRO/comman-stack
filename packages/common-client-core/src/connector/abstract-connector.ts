import { IFeature, FeatureParams, ClientStateParams } from '../interfaces';
import { merge, map, union, without, castArray } from 'lodash';

const combine = (features, extractor) => without(union(...map(features, res => castArray(extractor(res)))), undefined);
export const featureCatalog = {};


export abstract class AbstractFeature implements IFeature {
    public link: any;
    public createFetch: any;
    public connectionParam: any;
    public reducer: any;
    public resolver: any;
    public routerFactory: any;
    public route: any;
    public routeConfig: any;
    public navItem: any;
    public navItemRight: any;
    public rootComponentFactory: any[];
    public dataRootComponent: any[];
    public createFetchOptions: any[];
    public stylesInsert: any[];
    public scriptsInsert: any[];
    public catalogInfo: any[];
    public languagesFuncs: any[];
    public data: any[];

    constructor(
        feature?: FeatureParams,
        ...features,
    ) {

        // Connectivity
        this.link = combine(arguments, arg => arg.link);
        this.createFetch = combine(arguments, arg => arg.createFetch)
            .slice(-1)
            .pop();
        this.connectionParam = combine(arguments, arg => arg.connectionParam);

        // State management
        this.reducer = combine(arguments, arg => arg.reducer);
        this.resolver = combine(arguments, arg => arg.resolver);

        // Navigation
        this.routerFactory = combine(arguments, arg => arg.routerFactory)
            .slice(-1)
            .pop();
        this.route = combine(arguments, arg => arg.route);
        this.navItem = combine(arguments, arg => arg.navItem);
        this.navItemRight = combine(arguments, arg => arg.navItemRight);
        this.routeConfig = combine(arguments, arg => arg.routeConfig);

        // UI provider-components
        this.rootComponentFactory = combine(arguments, arg => arg.rootComponentFactory);
        this.dataRootComponent = combine(arguments, arg => arg.dataRootComponent);

        // UI provider-components
        this.languagesFuncs = combine(arguments, arg => arg.languagesFuncs);


        // TODO: Use React Helmet for those. Low level DOM manipulation
        this.stylesInsert = combine(arguments, arg => arg.stylesInsert);
        this.scriptsInsert = combine(arguments, arg => arg.scriptsInsert);

        // Shared modules data
        this.data = combine([{}].concat(Array.from(arguments)), arg => arg.data)
            .reduce(
                (acc, el) => [{ ...acc[0], ...el }],
                [{}],
        );
    }

    get router() {
        return this.routerFactory(this.routes);
    }

    public abstract get routes();

    public abstract get configuredRoutes();

    public abstract get navItems();

    public abstract get navItemsRight();

    get reducers() {
        return merge(...this.reducer);
    }

    get resolvers() {
        return merge(...this.resolver);
    }


    get connectionParams() {
        return this.connectionParam;
    }

    get stylesInserts() {
        return this.stylesInsert;
    }

    get scriptsInserts() {
        return this.scriptsInsert;
    }

    public abstract getWrappedRoot(root, req);

    public abstract getDataRoot(root);

    public abstract registerLanguages(monaco);
}
