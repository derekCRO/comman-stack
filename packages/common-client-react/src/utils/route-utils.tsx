import * as React from 'react';
import * as Loadable from 'react-loadable';
import { IRouteData } from '../interfaces';
import { matchRoutes } from 'react-router-config';
export const dynamicWrapper = (component: () => any, loading?: any) => Loadable({
    loader: component,
    loading: loading || <div> Loading...</div>,
});




export function getRelation(str1: string, str2: string) {
    if (str1 === str2) {
        console.warn('Two paths are equal');
    }
    const arr1 = str1.split('/');
    const arr2 = str2.split('/');
    if (arr2.every((item, index) => item === arr1[index])) {
        return 1;
    } else if (arr1.every((item, index) => item === arr2[index])) {
        return 2;
    }
    return 3;
}

export function getRenderArr(routes: string[]) {
    let renderArr = [];
    renderArr.push(routes[0]);
    for (let i = 1; i < routes.length; i += 1) {
        let isAdd = false;
        isAdd = renderArr.every(item => getRelation(item, routes[i]) === 3);
        renderArr = renderArr.filter(item => getRelation(item, routes[i]) !== 1);
        if (isAdd) {
            renderArr.push(routes[i]);
        }
    }
    return renderArr;
}

/**
 * Provides the routes based on the index search path.
 * For example, for routerData = {
 * ['/a']: {},
 * ['/a/1]: {},
 * ['/a/2]: {},
 * ['/a/2/1]: {},
 * ['/b/1]: {},
 * ['/b/2]: {},
 * ['/ab/1]: {},
 * }
 *
 * result of getRoutes('/a', routerData) will be
 * [
 * {path: '/a/1', ...},
 * {path: '/a/2', ...},
 * ]
 *
 *
 * @param path
 * @param routerData
 */
export function getRoutes(routes:string, location:string) {
    const branch = matchRoutes(routes, location.pathname)
  
    const promises = branch.map(({ route, match }) => {
      return route.loadData
        ? route.loadData(match)
        : Promise.resolve(null)
    })
  
    return Promise.all(promises)
}


 
