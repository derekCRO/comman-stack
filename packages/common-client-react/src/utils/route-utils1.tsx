import * as React from 'react';
import * as Loadable from 'react-loadable';
import { IRouteData } from '../interfaces';

export const dynamicWrapper = (component: () => any, loading?: any) => Loadable({
    loader: component,
    loading: loading || <div> Loading...</div>,
});




export function getRelation(str1: string, str2: string) {
    if (str1 === str2) {
        console.warn('Two paths are equal');
    }
    const arr1 = str1.split('/');
    const arr2 = str2.split('/');
    if (arr2.every((item, index) => item === arr1[index])) {
        return 1;
    } else if (arr1.every((item, index) => item === arr2[index])) {
        return 2;
    }
    return 3;
}

export function getRenderArr(routes: string[]) {
    let renderArr = [];
    renderArr.push(routes[0]);
    for (let i = 1; i < routes.length; i += 1) {
        let isAdd = false;
        isAdd = renderArr.every(item => getRelation(item, routes[i]) === 3);
        renderArr = renderArr.filter(item => getRelation(item, routes[i]) !== 1);
        if (isAdd) {
            renderArr.push(routes[i]);
        }
    }
    return renderArr;
}

/**
 * Provides the routes based on the index search path.
 * For example, for routerData = {
 * ['/a']: {},
 * ['/a/1]: {},
 * ['/a/2]: {},
 * ['/a/2/1]: {},
 * ['/b/1]: {},
 * ['/b/2]: {},
 * ['/ab/1]: {},
 * }
 *
 * result of getRoutes('/a', routerData) will be
 * [
 * {path: '/a/1', ...},
 * {path: '/a/2', ...},
 * ]
 *
 *
 * @param path
 * @param routerData
 */
export function getRoutes(path: string, routerData: IRouteData) {
    if (path[path.length - 1] !== '/') {
        path += '/'; //  Add a '/' to exclude incomplete paths
    }
    let routes = Object.keys(routerData).filter(routePath => {
        return routePath.indexOf(path) === 0 && routePath !== path;
    });
    // Replace path to '' eg. path='user' /user/name => name.
    routes = routes.map(item => item.replace(path, ''));
    // Get the route to be rendered to remove the deep rendering.
    // const renderArr = getRenderArr(routes);
    // Conversion and stitching parameters.
    const renderRoutes = routes.map(item => {
        const exact = !routes.some(route => route !== item && getRelation(route, item) === 1);
        const routeObject = { ...routerData[`${path}${item}`] };
        return {
            ...routeObject,
            key: `${path}${item}`,
            path: `${path}${item}`,
            // component: dynamicWrapper(routeObject.component, routeObject.loading),
            exact,
        };
    });
    return renderRoutes;
}
