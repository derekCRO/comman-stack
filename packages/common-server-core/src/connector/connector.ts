import { IResolverOptions, IDirectiveOptions } from '../interfaces';
import { merge, map, union, without, castArray } from 'lodash';

export const featureCatalog: any = {};

const combine = (features, extractor): any =>
  without(union(...map(features, res => castArray(extractor(res)))), undefined);

export type FeatureParams = {
  schema?: string | string[],
  createRemoteSchemas?: Function | Function[],
  createDirectivesFunc?: Function | Function[],
  createResolversFunc?: Function | Function[],
  createContextFunc?: Function | Function[],
  beforeware?: any | any[],
  middleware?: any | any[],
  catalogInfo?: any | any[],
};

class Feature {
  public schema: string[];
  public createRemoteSchemas?: Function | Function[];
  public createDirectivesFunc: Function[];
  public createResolversFunc: Function[];
  public createContextFunc: Function[];
  public beforeware: Function[];
  public middleware: Function[];

  constructor(feature?: FeatureParams, ...features: Feature[]) {
    // console.log(feature.schema[0] instanceof string);
    combine(arguments, arg => arg.catalogInfo).forEach(info =>
      Object.keys(info).forEach(key => (featureCatalog[key] = info[key])),
    );
    this.schema = combine(arguments, arg => arg.schema);
    this.createDirectivesFunc = combine(arguments, arg => arg.createDirectivesFunc);
    this.createResolversFunc = combine(arguments, arg => arg.createResolversFunc);
    this.createContextFunc = combine(arguments, arg => arg.createContextFunc);
    this.beforeware = combine(arguments, arg => arg.beforeware);
    this.middleware = combine(arguments, arg => arg.middleware);
  }

  get schemas(): string[] {
    return this.schema;
  }

  public async createContext(req: any, connectionParams: any, webSocket?: any) {
    const results = await Promise.all(
      this.createContextFunc.map(createContext => createContext(req, connectionParams, webSocket)),
    );
    return merge({}, ...results);
  }

  public createResolvers(options?: IResolverOptions) {
    return merge({}, ...this.createResolversFunc.map(createResolvers => createResolvers(options)));
  }

  public createDirectives(options?: IDirectiveOptions) {
    return merge({}, ...this.createDirectivesFunc.map(createDirectives => createDirectives(options)));
  }

  get beforewares(): any[] {
    return this.beforeware;
  }

  get middlewares(): any[] {
    return this.middleware;
  }
}

export { Feature };
