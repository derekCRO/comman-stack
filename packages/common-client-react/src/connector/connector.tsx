import * as React from 'react';
import { Route } from 'react-router-dom';
import { AbstractFeature, IFeature } from '@common-stack/client-core';
import { getRoutes } from '../utils';
import { renderRoutes } from 'react-router-config';

export class Feature extends AbstractFeature implements IFeature {
    /**
     * Get the routes
     */
    get routes() {
        const configuredRoutes = this.configuredRoutes;

        return [
            ...this.route.map((component: React.ReactElement<any>, idx: number) =>
                React.cloneElement(component, { key: idx + this.route.length }),
            ), renderRoutes(configuredRoutes),
        ];
    }

    /**
     * get configured routes.
     * Note: It overwrites the any duplicate key with latest loaded key.
     * TODO: Find a way to warn when there are duplicate keys.
     */
    get configuredRoutes() {
        const routes = Object.assign({}, ...this.routeConfig);
        return getRoutes('', { ...routes });
    }

    get navItems() {
        return this.navItem.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    get navItemsRight() {
        return this.navItemRight.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    public getWrappedRoot(root: React.ReactNode, req?: any): React.ReactNode {
        let nestedRoot = root;
        for (const componentFactory of this.rootComponentFactory) {
            nestedRoot = React.cloneElement(componentFactory(req), {}, nestedRoot);
        }
        return nestedRoot;
    }

    public getDataRoot(root) {
        let nestedRoot = root;
        for (const component of this.dataRootComponent) {
            nestedRoot = React.createElement(component, {}, nestedRoot);
        }
        return nestedRoot;
    }

    public registerLanguages(monaco) {
        for (const func of this.languagesFuncs) {
            func(monaco);
        }
    }
}
