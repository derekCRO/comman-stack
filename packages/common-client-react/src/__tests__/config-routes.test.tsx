import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as Loadable from 'react-loadable';
import { Route } from 'react-router-dom';
import { getRoutes } from '../utils';
import { Feature } from '../connector';
import { IRouteData } from '../interfaces';
import 'jest';


function waitFor(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}

function createLoader(delay, loader, error?: any) {
    return () => {
        return waitFor(delay).then(() => {
            if (loader) {
                return loader();
            } else {
                throw error;
            }
        });
    };
}

function MyLoadingComponent(props) {
    return <div>MyLoadingComponent {JSON.stringify(props)}</div>;
}

function MyComponent(props) {
    return <div>MyComponent {JSON.stringify(props)}</div>;
}


afterEach(async () => {
    try {
        await Loadable.preloadAll();
    } catch (err) { }
});

// Test case borrowed from
// https://github.com/jamiebuilds/react-loadable/blob/master/__tests__/test.js
test('loading success', async () => {
    let LoadableMyComponent = Loadable<{ prop: any }, {}>({
        loader: createLoader(400, () => MyComponent),
        loading: MyLoadingComponent,
    });

    let component1 = renderer.create(<LoadableMyComponent prop="foo" />);

    expect(component1.toJSON()).toMatchSnapshot(); // initial
    await waitFor(200);
    expect(component1.toJSON()).toMatchSnapshot(); // loading
    await waitFor(200);
    expect(component1.toJSON()).toMatchSnapshot(); // loaded

    let component2 = renderer.create(<LoadableMyComponent prop="bar" />);

    expect(component2.toJSON()).toMatchSnapshot(); // reload
});

const routerConfig = (namespace = '') => ({
    [namespace + '/a']: {
        component: () => MyComponent,
    },
    ['/a/1']: {
        component: () => MyComponent,
    },
    [namespace + '/a/1']: {
        component: () => MyComponent,
    },
    [namespace + '/a/2']: {
        component: () => MyComponent,
    },
    [namespace + '/a/2/1']: {
        component: () => MyComponent,
    },
    [namespace + '/ab/2/1']: {
        component: () => MyComponent,
    },
    [namespace + '/b/1']: {
        component: () => MyComponent,
    },
    [namespace + '/b/login/register']: {
        component: () => MyComponent,
    },
});

describe('routeConfig getRoutes', () => {


    test('getRoutes based on index path', async () => {
        const result = ['/a/1', '/a/2', '/a/2/1'];
        expect(getRoutes('/a', routerConfig()).map(item => item.path)).toEqual(result);
    });

    test('getRoutes based on index path', async () => {

        const result = ['/a', '/a/1', '/a/2', '/a/2/1', '/ab/2/1', '/b/1', '/b/login/register'];
        expect(getRoutes('', routerConfig()).map(item => item.path)).toEqual(result);
    });

    test('getRoutes with `@`in namespace', async () => {

        const result = ['@namespace/a', '@namespace/a/1',
            '@namespace/a/2', '@namespace/a/2/1', '@namespace/ab/2/1', '@namespace/b/1', '@namespace/b/login/register'];
        expect(getRoutes('@namespace', routerConfig('@namespace')).map(item => item.path)).toEqual(result);
    });
});



describe('connector configuredRoutes', () => {

    test('getRoutes based on index path', async () => {

        const connector = new Feature({ routeConfig: routerConfig() });
        const result = [{ key: '/a', path: '/a', exact: false },
        { key: '/a/1', path: '/a/1', exact: true },
        { key: '/a/2', path: '/a/2', exact: false },
        {
            key: '/a/2/1',
            path: '/a/2/1',
            exact: true,
        },
        {
            key: '/ab/2/1',
            path: '/ab/2/1',
            exact: true,
        },
        { key: '/b/1', path: '/b/1', exact: true },
        {
            key: '/b/login/register',
            path: '/b/login/register',
            exact: true,
        }];

        expect(JSON.parse(JSON.stringify(connector.configuredRoutes))).toMatchObject(result);
    });

    test('getRoutes based without any namespace', async () => {

        const connector = new Feature({ routeConfig: routerConfig('@namespace') });
        const result = [{ 'exact': true, 'key': '/a/1', 'path': '/a/1' }];

        expect(JSON.parse(JSON.stringify(connector.configuredRoutes))).toMatchObject(result);
    });
});


describe('connector routes', () => {
    const staticRoutes = {
        route: [
            <Route key={'static1'} exact={true} path="/static1" component={MyComponent} />,
            <Route key={'static2'} exact={true} path="/static2" component={MyComponent} />,
        ],
    };

    test('check static routes', async () => {
        const connector = new Feature(staticRoutes);

        expect(connector.routes).toMatchSnapshot();
    });

    test('merge static routes and configurable routes', async () => {
        const connector = new Feature({ routeConfig: routerConfig() }, new Feature(staticRoutes));

        expect(connector.routes).toMatchSnapshot();
    });


    test('check configurable routes', async () => {
        const connector = new Feature({ routeConfig: routerConfig() });

        expect(connector.routes).toMatchSnapshot();
    });

});
