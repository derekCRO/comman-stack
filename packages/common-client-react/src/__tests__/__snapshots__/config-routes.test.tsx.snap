// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connector routes check configurable routes 1`] = `
Array [
  <Route
    component={[Function]}
    exact={false}
    path="/a"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/a/1"
  />,
  <Route
    component={[Function]}
    exact={false}
    path="/a/2"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/a/2/1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/ab/2/1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/b/1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/b/login/register"
  />,
]
`;

exports[`connector routes check static routes 1`] = `
Array [
  <Route
    component={[Function]}
    exact={true}
    path="/static1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/static2"
  />,
]
`;

exports[`connector routes getRoutes based on index path 1`] = `
Array [
  <Route
    component={[Function]}
    exact={true}
    path="/static1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/static2"
  />,
  <Route
    component={[Function]}
    exact={false}
    path="/a"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/a/1"
  />,
  <Route
    component={[Function]}
    exact={false}
    path="/a/2"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/a/2/1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/ab/2/1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/b/1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/b/login/register"
  />,
]
`;

exports[`connector routes merge static routes and configurable routes 1`] = `
Array [
  <Route
    component={[Function]}
    exact={true}
    path="/static1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/static2"
  />,
  <Route
    component={[Function]}
    exact={false}
    path="/a"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/a/1"
  />,
  <Route
    component={[Function]}
    exact={false}
    path="/a/2"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/a/2/1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/ab/2/1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/b/1"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/b/login/register"
  />,
]
`;

exports[`loading success 1`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":false,"timedOut":false,"error":null}
</div>
`;

exports[`loading success 2`] = `
<div>
  MyLoadingComponent 
  {"isLoading":true,"pastDelay":true,"timedOut":false,"error":null}
</div>
`;

exports[`loading success 3`] = `
<div>
  MyComponent 
  {"prop":"foo"}
</div>
`;

exports[`loading success 4`] = `
<div>
  MyComponent 
  {"prop":"bar"}
</div>
`;
